aten/src/ATen/core/Half.h

```diff
@@ -74,18 +76,55 @@ struct alignas(4) ComplexHalf {
   Half real_;
   Half imag_;
   ComplexHalf() = default;
+  Half real() const { return real_; }
+  Half imag() const { return imag_; }
+  inline AT_HOSTDEVICE ComplexHalf(std::complex<float> value)
+    : real_(value.real()), imag_(value.imag()) {}
+  inline AT_HOSTDEVICE operator std::complex<float>() const {
+    return {real_, imag_};
+  }
 };
 
+template <typename T>
+struct is_complex_t : public std::false_type {};
+
+template <typename T>
+struct is_complex_t<std::complex<T>> : public std::true_type {};
+
+template <>
+struct is_complex_t<ComplexHalf> : public std::true_type {};
+
+// Extract double from std::complex<double>; is identity otherwise
+// TODO: Write in more idiomatic C++17
+template <typename T> struct scalar_value_type                  { typedef T type; };
+template <typename T> struct scalar_value_type<std::complex<T>> { typedef T type; };
+template <>           struct scalar_value_type<ComplexHalf>     { typedef Half type; };
+
 template <typename To, typename From>
-To convert(From f) {
+typename std::enable_if<
+  c10::guts::negation<
+    c10::guts::conjunction<
+      is_complex_t<From>,
+      c10::guts::negation<is_complex_t<To>>
+    >
+  >::value, To>::type convert(From f) {
   return static_cast<To>(f);
 }
 
+template <typename To, typename From>
+typename std::enable_if<
+  c10::guts::conjunction<
+    is_complex_t<From>,
+    c10::guts::negation<is_complex_t<To>>
+  >::value, To>::type convert(From f) {
+  return static_cast<To>(f.real());
```

Comment:
don't you want to call convert here, not static_cast?  What happens if you try to go complex(Float,Double) to/from Half?

Reply:
static cast is fine. You'll do a static cast from float to at::Half and that's well defined (because at::Half defines appropriate implicit conversions).